lib/linked_list.rb

```diff
@@ -18,71 +19,161 @@ def initialize
 
     # method to add a new node with the specific data value in the linked list
     # insert the new node at the beginning of the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def add_first(value)
-      raise NotImplementedError
-    end
+      if @head.nil?
+        @head = Node.new(value)
+      else 
+        new_nodelet = Node.new(value, @head)
+        new_nodelet.next = @head 
+        @head = new_nodelet 
+      end 
+    end 
 
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def search(value)
-      raise NotImplementedError
+      return false if @head.nil?
+
+      current = @head
+
+      while current != nil 
+        return true if value == current.data 
+        current = current.next
+      end 
+
+      return false 
     end
 
     # method to return the max value in the linked list
     # returns the data value and not the node
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def find_max

```

Comment:
Correct, but space complexity would be O(1) since you don't create a new data structure, but just track the greatest value found so far.